{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Solution:**\n",
    "\n",
    "First, we want to combine the 9 different Excel spreadsheets into one data frame. We do this by for-loop: first, outside the loop, we read the first file (MID1.xls) to a data frame (rna_data), and then loop over numbers 2-9 and read and append each file's contents to our data frame. In order to preserve information on which entry came from which file, we create a `MultiIndex` index, where the first level is the file name and the second level is the index of the entry (row number, starting from 0). Column names in the data frame default to 0 and 1, and we leave them as is.\n",
    "\n",
    "Once the data has been read, we name the first level of the `MultiIndex` \"BARCODE\" and the second level \"index_no\". Then, we read the metadata file (metadata.xls) to a data frame, and join it with the rna_data data frame. The join is performed on the common index, which in this case is \"BARCODE\". Since all \"BARCODE\" values are contained in both data frames, it doesn't matter whether we do an inner or outer join.\n",
    "\n",
    "Finally, we change NaN values to \"unknown\", and check that our indices are unique and that no NaN values are left in the data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "# Load first file\n",
    "rna_data = pd.read_excel(DATA_FOLDER+\"/microbiome/MID1.xls\",sheetname=\"Sheet 1\",header=None)\n",
    "rna_indices = [[\"MID1\"]*len(rna_data.index), list(rna_data.index)]\n",
    "#rna_indices = [[\"MID1\"]*len(rna_data.index), list(rna_data.index.astype(str))]\n",
    "tuples = list(zip(*rna_indices))\n",
    "rna_data.index = pd.MultiIndex.from_tuples(tuples)\n",
    "\n",
    "#rna_data.index = [[\"MID1_\"]*len(rna_data.index), list(rna_data.index.astype(str))]\n",
    "\n",
    "# Append other files\n",
    "for ii in range(2,10):\n",
    "    next_sheet = pd.read_excel(DATA_FOLDER+\"/microbiome/MID%d.xls\" % ii,sheetname=\"Sheet 1\",header=None)\n",
    "    rna_indices = [[\"MID%d\" % ii]*len(next_sheet.index), list(next_sheet.index)]\n",
    "    tuples = list(zip(*rna_indices))\n",
    "    next_sheet.index = pd.MultiIndex.from_tuples(tuples)\n",
    "    #next_sheet.index = \"MID%d_\" % ii + next_sheet.index.astype(str)\n",
    "    rna_data = rna_data.append(next_sheet,verify_integrity=False)\n",
    "\n",
    "# Change index names to match metadata\n",
    "rna_data.index.names = [\"BARCODE\",\"index_no\"]\n",
    "\n",
    "# Read and join metadata\n",
    "rna_metadata = pd.read_excel(DATA_FOLDER+\"/microbiome/metadata.xls\",index_col=\"BARCODE\")\n",
    "#pd.merge(rna_data,rna_metadata,left_index=True,right_on=\"BARCODE\")\n",
    "#rna_metadata.join(rna_data)\n",
    "rna_data = rna_data.join(rna_metadata, how='outer')\n",
    "\n",
    "print(rna_data.isnull().any().any())\n",
    "\n",
    "# Change NaNs to unknown\n",
    "rna_data.fillna(value=\"unknown\",inplace=True)\n",
    "# Check that no NaNs are left\n",
    "print(rna_data.isnull().any().any())\n",
    "\n",
    "#rna_data = rna_data.append(pd.read_excel(DATA_FOLDER+\"/microbiome/MID2.xls\",sheetname=\"Sheet 1\",header=None))\n",
    "#rna_data\n",
    "#print(rna_metadata.index)\n",
    "#rna_data.index.is_unique\n",
    "#rna_data.loc[\"MID5\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>index_no</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID1</th>\n",
       "      <th>0</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Archaea \"Crenarchaeota\" Thermoprotei Thermopro...</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Halobacteria Halobacte...</td>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanococci Methanoco...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Methanopyri Methanopyr...</td>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Archaea \"Euryarchaeota\" Thermoplasmata Thermop...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Bacteria \"Actinobacteria\" Actinobacteria Actin...</td>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID9</th>\n",
       "      <th>251</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>805</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>862</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3534</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>25</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>Bacteria \"Proteobacteria\" Gammaproteobacteria ...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>Bacteria \"Synergistetes\" Synergistia Synergist...</td>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>Bacteria \"Tenericutes\" Mollicutes Anaeroplasma...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>Bacteria \"Tenericutes\" Mollicutes Haloplasmata...</td>\n",
       "      <td>11</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>Bacteria \"Thermodesulfobacteria\" Thermodesulfo...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>Bacteria \"Thermodesulfobacteria\" Thermodesulfo...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>Bacteria \"Thermotogae\" Thermotogae Thermotogal...</td>\n",
       "      <td>43</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Bacteria \"Thermotogae\" Thermotogae Thermotogal...</td>\n",
       "      <td>16</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>Bacteria \"Verrucomicrobia\" Opitutae Opitutales...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Bacteria Cyanobacteria Cyanobacteria  Chloropl...</td>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  0     1  \\\n",
       "BARCODE index_no                                                            \n",
       "MID1    0         Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...     7   \n",
       "        1         Archaea \"Crenarchaeota\" Thermoprotei Desulfuro...     2   \n",
       "        2         Archaea \"Crenarchaeota\" Thermoprotei Sulfoloba...     3   \n",
       "        3         Archaea \"Crenarchaeota\" Thermoprotei Thermopro...     3   \n",
       "        4         Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...     7   \n",
       "        5         Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...     1   \n",
       "        6         Archaea \"Euryarchaeota\" \"Methanomicrobia\" Meth...     1   \n",
       "        7         Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...     1   \n",
       "        8         Archaea \"Euryarchaeota\" Archaeoglobi Archaeogl...     1   \n",
       "        9         Archaea \"Euryarchaeota\" Halobacteria Halobacte...     1   \n",
       "        10        Archaea \"Euryarchaeota\" Halobacteria Halobacte...     4   \n",
       "        11        Archaea \"Euryarchaeota\" Halobacteria Halobacte...     1   \n",
       "        12        Archaea \"Euryarchaeota\" Halobacteria Halobacte...     2   \n",
       "        13        Archaea \"Euryarchaeota\" Halobacteria Halobacte...     4   \n",
       "        14        Archaea \"Euryarchaeota\" Methanococci Methanoco...     1   \n",
       "        15        Archaea \"Euryarchaeota\" Methanopyri Methanopyr...    12   \n",
       "        16        Archaea \"Euryarchaeota\" Thermoplasmata Thermop...     2   \n",
       "        17        Bacteria \"Actinobacteria\" Actinobacteria Actin...     2   \n",
       "        18        Bacteria \"Actinobacteria\" Actinobacteria Actin...     1   \n",
       "        19        Bacteria \"Actinobacteria\" Actinobacteria Actin...     1   \n",
       "        20        Bacteria \"Actinobacteria\" Actinobacteria Actin...     1   \n",
       "        21        Bacteria \"Actinobacteria\" Actinobacteria Actin...     2   \n",
       "        22        Bacteria \"Actinobacteria\" Actinobacteria Actin...    53   \n",
       "        23        Bacteria \"Actinobacteria\" Actinobacteria Actin...     1   \n",
       "        24        Bacteria \"Actinobacteria\" Actinobacteria Actin...     2   \n",
       "        25        Bacteria \"Actinobacteria\" Actinobacteria Actin...     3   \n",
       "        26        Bacteria \"Actinobacteria\" Actinobacteria Actin...     1   \n",
       "        27        Bacteria \"Actinobacteria\" Actinobacteria Actin...    99   \n",
       "        28        Bacteria \"Actinobacteria\" Actinobacteria Actin...    17   \n",
       "        29        Bacteria \"Actinobacteria\" Actinobacteria Actin...     2   \n",
       "...                                                             ...   ...   \n",
       "MID9    251       Bacteria \"Proteobacteria\" Gammaproteobacteria ...    10   \n",
       "        252       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        253       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        254       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        255       Bacteria \"Proteobacteria\" Gammaproteobacteria ...    12   \n",
       "        256       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     4   \n",
       "        257       Bacteria \"Proteobacteria\" Gammaproteobacteria ...   805   \n",
       "        258       Bacteria \"Proteobacteria\" Gammaproteobacteria ...   862   \n",
       "        259       Bacteria \"Proteobacteria\" Gammaproteobacteria ...    12   \n",
       "        260       Bacteria \"Proteobacteria\" Gammaproteobacteria ...  3534   \n",
       "        261       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     4   \n",
       "        262       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        263       Bacteria \"Proteobacteria\" Gammaproteobacteria ...    25   \n",
       "        264       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     2   \n",
       "        265       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     3   \n",
       "        266       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     2   \n",
       "        267       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        268       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     3   \n",
       "        269       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     2   \n",
       "        270       Bacteria \"Proteobacteria\" Gammaproteobacteria ...     1   \n",
       "        271       Bacteria \"Synergistetes\" Synergistia Synergist...     3   \n",
       "        272       Bacteria \"Tenericutes\" Mollicutes Anaeroplasma...     1   \n",
       "        273       Bacteria \"Tenericutes\" Mollicutes Haloplasmata...    11   \n",
       "        274       Bacteria \"Thermodesulfobacteria\" Thermodesulfo...     1   \n",
       "        275       Bacteria \"Thermodesulfobacteria\" Thermodesulfo...     1   \n",
       "        276       Bacteria \"Thermotogae\" Thermotogae Thermotogal...    43   \n",
       "        277       Bacteria \"Thermotogae\" Thermotogae Thermotogal...    16   \n",
       "        278       Bacteria \"Verrucomicrobia\" Opitutae Opitutales...     1   \n",
       "        279       Bacteria Cyanobacteria Cyanobacteria  Chloropl...     1   \n",
       "        280       Bacteria Cyanobacteria Cyanobacteria  Chloropl...    10   \n",
       "\n",
       "                               GROUP   SAMPLE  \n",
       "BARCODE index_no                               \n",
       "MID1    0         EXTRACTION CONTROL  unknown  \n",
       "        1         EXTRACTION CONTROL  unknown  \n",
       "        2         EXTRACTION CONTROL  unknown  \n",
       "        3         EXTRACTION CONTROL  unknown  \n",
       "        4         EXTRACTION CONTROL  unknown  \n",
       "        5         EXTRACTION CONTROL  unknown  \n",
       "        6         EXTRACTION CONTROL  unknown  \n",
       "        7         EXTRACTION CONTROL  unknown  \n",
       "        8         EXTRACTION CONTROL  unknown  \n",
       "        9         EXTRACTION CONTROL  unknown  \n",
       "        10        EXTRACTION CONTROL  unknown  \n",
       "        11        EXTRACTION CONTROL  unknown  \n",
       "        12        EXTRACTION CONTROL  unknown  \n",
       "        13        EXTRACTION CONTROL  unknown  \n",
       "        14        EXTRACTION CONTROL  unknown  \n",
       "        15        EXTRACTION CONTROL  unknown  \n",
       "        16        EXTRACTION CONTROL  unknown  \n",
       "        17        EXTRACTION CONTROL  unknown  \n",
       "        18        EXTRACTION CONTROL  unknown  \n",
       "        19        EXTRACTION CONTROL  unknown  \n",
       "        20        EXTRACTION CONTROL  unknown  \n",
       "        21        EXTRACTION CONTROL  unknown  \n",
       "        22        EXTRACTION CONTROL  unknown  \n",
       "        23        EXTRACTION CONTROL  unknown  \n",
       "        24        EXTRACTION CONTROL  unknown  \n",
       "        25        EXTRACTION CONTROL  unknown  \n",
       "        26        EXTRACTION CONTROL  unknown  \n",
       "        27        EXTRACTION CONTROL  unknown  \n",
       "        28        EXTRACTION CONTROL  unknown  \n",
       "        29        EXTRACTION CONTROL  unknown  \n",
       "...                              ...      ...  \n",
       "MID9    251                Control 2    stool  \n",
       "        252                Control 2    stool  \n",
       "        253                Control 2    stool  \n",
       "        254                Control 2    stool  \n",
       "        255                Control 2    stool  \n",
       "        256                Control 2    stool  \n",
       "        257                Control 2    stool  \n",
       "        258                Control 2    stool  \n",
       "        259                Control 2    stool  \n",
       "        260                Control 2    stool  \n",
       "        261                Control 2    stool  \n",
       "        262                Control 2    stool  \n",
       "        263                Control 2    stool  \n",
       "        264                Control 2    stool  \n",
       "        265                Control 2    stool  \n",
       "        266                Control 2    stool  \n",
       "        267                Control 2    stool  \n",
       "        268                Control 2    stool  \n",
       "        269                Control 2    stool  \n",
       "        270                Control 2    stool  \n",
       "        271                Control 2    stool  \n",
       "        272                Control 2    stool  \n",
       "        273                Control 2    stool  \n",
       "        274                Control 2    stool  \n",
       "        275                Control 2    stool  \n",
       "        276                Control 2    stool  \n",
       "        277                Control 2    stool  \n",
       "        278                Control 2    stool  \n",
       "        279                Control 2    stool  \n",
       "        280                Control 2    stool  \n",
       "\n",
       "[2396 rows x 4 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
